File: problems/problemset/rate-limiter/C++/runner/Testbench.cpp
#include <bits/stdc++.h>
#include <nlohmann/json.hpp>
#include <chrono>
#include <sys/resource.h>
#include "../code/RateLimiter.h"

using json = nlohmann::json;

static long peakMemoryKB() {
    struct rusage usage;
    getrusage(RUSAGE_SELF, &usage);
    return usage.ru_maxrss;
}

int main() {
    std::string testcase_str((std::istreambuf_iterator<char>(std::cin)),
                              std::istreambuf_iterator<char>());

    json testcase = json::parse(testcase_str);

    auto input = testcase["input"].get<std::string>();
    std::stringstream ss(input);
    std::string cmd;

    std::string output;
    RateLimiter* rl = nullptr;

    auto start = std::chrono::high_resolution_clock::now();

    while (ss >> cmd) {
        if (cmd == "INIT") {
            int maxRequests;
            long long windowMs;
            ss >> maxRequests >> windowMs;
            delete rl;
            rl = new RateLimiter(maxRequests, windowMs);
        } else if (cmd == "REQUEST") {
            long long timestamp;
            ss >> timestamp;
            output += (rl->allow(timestamp) ? "ALLOWED" : "DENIED");
            output += "\n";
        }
    }

    auto end = std::chrono::high_resolution_clock::now();
    double time_ms   = std::chrono::duration<double, std::milli>(end - start).count();
    long   memory_kb = peakMemoryKB();

    delete rl;

    json result;
    result["actual_output"] = output;
    result["time_ms"]       = time_ms;
    result["memory_kb"]     = memory_kb;

    std::cout << "__BEGIN_RESULT__" << result.dump() << "__END_RESULT__";
}

File: problems/problemset/rate-limiter/Python/runner/testbench.py
import sys
import json
import time
import resource

sys.path.insert(0, '/workspace/code')
from rate_limiter import RateLimiter

def main():
    testcase = json.loads(sys.stdin.read())

    rl = None
    output_lines = []

    start = time.perf_counter()

    for line in testcase['input'].splitlines():
        parts = line.split()
        if not parts:
            continue
        cmd = parts[0]
        if cmd == 'INIT':
            rl = RateLimiter(int(parts[1]), int(parts[2]))
        elif cmd == 'REQUEST':
            output_lines.append('ALLOWED' if rl.allow(int(parts[1])) else 'DENIED')

    elapsed_ms = (time.perf_counter() - start) * 1000
    memory_kb = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss

    actual_output = '\n'.join(output_lines) + '\n' if output_lines else ''

    result = {
        'actual_output': actual_output,
        'time_ms': elapsed_ms,
        'memory_kb': memory_kb,
    }

    sys.stdout.write(f'__BEGIN_RESULT__{json.dumps(result)}__END_RESULT__')

main()
